### PR TITLE
cosocket: add function `tcpsock:setclientcert`, reimplemented `tcpsock:sslhandshake` with FFI.

### PR DESCRIPTION
**Pull request by dndx**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

Sister PR: https://github.com/openresty/lua-resty-core/pull/278

### CODE REVIEW

File: src/ngx_http_lua_socket_tcp.c [hunk 1]
```diff
@@ -1980,12 +1960,14 @@ ngx_http_lua_socket_prepare_error_retvals(ngx_http_request_t *r,
...
 
-    if (ft_type & (NGX_HTTP_LUA_SOCKET_FT_RESOLVER
-                   | NGX_HTTP_LUA_SOCKET_FT_SSL))
-    {
+    if (ft_type & NGX_HTTP_LUA_SOCKET_FT_RESOLVER) {
         return 2;
     }
 
+    if (ft_type & NGX_HTTP_LUA_SOCKET_FT_SSL) {
+        return 0;
```
Comment (kikito) [inline src/ngx_http_lua_socket_tcp.c]: Unclear about what the return values for this function mean. Would it make sense to replace the `0` and the `2` by names, like `NGX_AGAIN` and `NGX_ERROR`?

Comment (dndx) [reply src/ngx_http_lua_socket_tcp.c]: This is the number of return values to the caller of the function. It is not a special constant.

File: t/162-socket-tls-handshake.t [hunk 2]
```diff
@@ -0,0 +1,439 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
```
Comment (thibaultcha) [inline t/162-socket-tls-handshake.t]: I am wondering if this new test suite belongs to lua-resty-core instead. After all, all FFI-based APIs are tested there, and not in lua-nginx-module.

Comment (dndx) [reply t/162-socket-tls-handshake.t]: Yes I thought about this initially, but seems that this clear split is not being followed by existing tests like https://github.com/openresty/lua-nginx-module/blob/master/t/154-semaphore.t.

But, I don't have strong feeling about where to keep the cases, if moving is a better in your opinion then let's move it.

**Comment by shtaif** [comment]
@dndx @kikito @thibaultcha thank you for the great work! ✌️ ✌️ 
May I kindly ask, would this feature make it possible to dynamically determine the TLS client certificate to use when proxying requests over to an upstream (on a per-request basis)?

**Comment by thibaultcha** [comment]
I just rebased this PR (and sibling PR) on top of the latest master branch now that OpenResty 1.17.8.1 has been released.

**Comment by dndx** [comment]
@shtaif This PR is for cosocket connections only. For setting client certs to use in upstream connections, https://github.com/Kong/lua-kong-nginx-module#restykongtlsset_upstream_cert_and_key can do it. You just need to apply the required patches from https://github.com/Kong/kong-build-tools/blob/master/openresty-patches/patches/1.15.8.3/nginx-1.15.8_01-upstream_client_certificate_and_ssl_verify.patch.

**Comment by shtaif** [comment]
> @shtaif This PR is for cosocket connections only. For setting client certs to use in upstream connections, https://github.com/Kong/lua-kong-nginx-module#restykongtlsset_upstream_cert_and_key can do it. You just need to apply the required patches from https://github.com/Kong/kong-build-tools/blob/master/openresty-patches/patches/1.15.8.3/nginx-1.15.8_01-upstream_client_certificate_and_ssl_verify.patch.

@dndx this was indeed the right way to go, thanks!

File: src/ngx_http_lua_socket_tcp.c [hunk 3]
```diff
@@ -1531,187 +1504,214 @@ ngx_http_lua_socket_conn_error_retval_handler(ngx_http_request_t *r,
...
-    ngx_ssl_session_t      **psession;
-    ngx_http_request_t      *r;
-    ngx_http_lua_ctx_t      *ctx;
-    ngx_http_lua_co_ctx_t   *coctx;
-
-    ngx_http_lua_socket_tcp_upstream_t  *u;
-
-    /* Lua function arguments: self [,session] [,host] [,verify]
-       [,send_status_req] */
+    if (ops & SOCKET_OP_CONNECT && u->conn_waiting) {
```
Comment (spacewander) [inline src/ngx_http_lua_socket_tcp.c:1511]: Would be better if the `xx & yy` is wrapped by `()` like similar code in this repo.

**Comment by spacewander** [comment]
@dndx 
Do you have time to solve the conflict? Thanks!

**Review by spacewander — APPROVED** [review]

**Review by rainingmaster — APPROVED** [review]

**Review by doujiang24 — APPROVED** [review]

**Comment by agentzh** [comment]
No, we never change existing API function name. They stay there forever. We take API backward compatibility very seriously.

**Comment by agentzh** [comment]
Also, table key lookups are slow which we've been trying hard to avoid in our core APIs.

**Comment by agentzh** [comment]
Also, we should add a separate method to load the client cert/key. So that when we re-connect, we don't have to load them all over again.

**Comment by astromechza** [comment]
@agentzh / @dndx do we have any updates on this? Is it in a merge-able state?

**Comment by pintsized** [comment]
@agentzh 

> No, we never change existing API function name. They stay there forever. We take API backward compatibility very seriously.
> Also, we should add a separate method to load the client cert/key. So that when we re-connect, we don't have to load them all over again.

If I understand you correctly, would it perhaps be better if we keep the `sslhandshake` signature as it is, and add something like a `sock:settlsclientcertificate(cert, key)` method?

I'm concerned about switching the name and signature of the handshake method, from the perspective of calling code. Ideally we could just supply an optional cert and private key to get mTLS support, without having to hold two different configurations around.

Are there any other broad concerns with this feature?

**Comment by agentzh** [comment]
@pintsized yeah, something like `sock:setclientcert(cert, key)`, but we should avoid parsing PEM-formatted certificates or keys upon every function invocation.

**Comment by dndx** [comment]
@agentzh I had a quick chat with @pintsized , in an interest to get this merged into mainline. Here is the proposal:

1. We implement a new Lua API `sock:setclientcert(cert, key)` using FFI, this accepts `cdata` object that corresponds to the parsed client cert and key
2. `sock:sslhandshake` signature stays the same and we will not attempt to reimplement it in FFI

Port over all the existing tests in this PR.

If we do the above, do you feel confident that this feature can be merged? We could implement the above functionality in a separate Nginx module, but @pintsized points out it will make supporting mTLS in `lua-resty-http` harder and make the OpenResty ecosystem more fragmented.

Thanks

File: README.markdown [hunk 4]
```diff
@@ -7738,6 +7739,31 @@ This method was first introduced in the `v0.5.0rc1` release.
 
 [Back to TOC](#nginx-api-for-lua)
 
+tcpsock:setclientcert
+--------------------
+
+**syntax:** *tcpsock:setclientcert(cert, pkey)*
```
Comment (pintsized) [inline README.markdown:7745]:
```suggestion
**syntax:** *ok, err = tcpsock:setclientcert(cert, pkey)*
```

**Review by pintsized — CHANGES_REQUESTED** [review]
Just a small suggestion for the docs, otherwise this is looking great to me 👍

File: README.markdown (hunk 4 above)
Comment (chronolaw) [reply README.markdown:7745]: sorry, I forgot it, I will update it soon, thanks.

**Comment by dndx** [comment]
@zhuizhuhaomeng @doujiang24 Could you please take another look at this updated PR with https://github.com/openresty/lua-resty-core/pull/278? It should now satisfy the API requirements from @agentzh above.

**Comment by dndx** [comment]
Pinging @zhuizhuhaomeng @spacewande @doujiang24 again.

**Comment by zhuizhuhaomeng** [comment]
> Pinging @zhuizhuhaomeng @spacewande @doujiang24 again.

I will review this PR when I have time.

File: src/ngx_http_lua_socket_tcp.c [hunk 5]
```diff
@@ -1559,187 +1534,214 @@ ngx_http_lua_socket_conn_error_retval_handler(ngx_http_request_t *r,
...
     if (u->request != r) {
-        return luaL_error(L, "bad request");
+        *errmsg = "bad request";
+        return NGX_ERROR;
     }
 
-    ngx_http_lua_socket_check_busy_connecting(r, u, L);
-    ngx_http_lua_socket_check_busy_reading(r, u, L);
-    ngx_http_lua_socket_check_busy_writing(r, u, L);
+    busy_rc = ngx_http_lua_socket_tcp_check_busy(r, u, SOCKET_OP_CONNECT
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply src/ngx_http_lua_socket_tcp.c]: done, thanks.

**Comment by zhuizhuhaomeng** [comment]
@chronolaw  would you please update the title to reflect the new changes.

**Review by zhuizhuhaomeng — APPROVED** [review]

**Comment by dndx** [comment]
@zhuizhuhaomeng the tile has been changed. We can squash the change if you prefer.

**Comment by zhuizhuhaomeng** [comment]
> 

You can squash the changes.

**Review by spacewander — APPROVED** [review]

**Comment by zhuizhuhaomeng** [comment]
@dndx  would you please add the script to generate the mtls certificates to t/cert.

**Review by doujiang24 — APPROVED** [review]

**Comment by dndx** [comment]
@zhuizhuhaomeng The script has been added. One of the Travis run failed because of "negative credit balance" for Travis. However, considering the other build did passed, I think the new script generated certs should be good:

https://app.travis-ci.com/github/openresty/lua-nginx-module/builds/247813268

**Comment by zhuizhuhaomeng** [comment]
merged with the following patch
```
--- a/.travis.yml
+++ b/.travis.yml
@@ -88,7 +88,7 @@ install:
   - git clone https://github.com/openresty/rds-json-nginx-module.git ../rds-json-nginx-module
   - git clone https://github.com/openresty/srcache-nginx-module.git ../srcache-nginx-module
   - git clone https://github.com/openresty/redis2-nginx-module.git ../redis2-nginx-module
-  - git clone -b feat/cosocket_tlshandshake https://github.com/dndx/lua-resty-core.git ../lua-resty-core
+  - git clone https://github.com/dndx/lua-resty-core.git ../lua-resty-core
   - git clone https://github.com/openresty/lua-resty-lrucache.git ../lua-resty-lrucache
   - git clone https://github.com/openresty/lua-resty-mysql.git ../lua-resty-mysql
   - git clone https://github.com/openresty/lua-resty-string.git ../lua-resty-string
```

**Comment by denissabramovs** [comment]
Where setclientcert method? Why this wasn't merged in the end?

**Pull request closed without merging**